[UXUI-155] Tag merge

andersonjeccel opened the pull request:
| Q                                      | A
| -------------------------------------- | --------------------------
| Bug fix?          | ❌
| New feature/enhancement?      | ✔️
| Deprecations?                          | ❌
| BC breaks?        | ❌
| Automated tests included?              | ✔️ 
| Related user documentation PR URL      | 
| Related developer documentation PR URL | 
| Issue(s) addressed                     | 

## Description

This PR adds tag merge functionality to Mautic, allowing users to merge one tag into another while maintaining data integrity and providing clear user feedback.

---
### 📋 Steps to test this PR:

1. Open this PR on Gitpod or pull down for testing locally
2. Navigate to Tags section in Mautic admin
3. Select a tag and click "Merge" action
4. Choose another tag to merge into
5. Verify the merge completes successfully
6. Check that the secondary tag is deleted
7. Verify all leads are properly associated with the primary tag
8. Confirm success flash message appears
9. Test that current tag is excluded from merge options


https://github.com/user-attachments/assets/deef7437-2af5-4468-ad39-26707d383c6b

andersonjeccel committed:
feat: add tag merge functionality with event system

andersonjeccel committed:
style: apply CS Fixer formatting to tag merge files

andersonjeccel committed:
test: add comprehensive test coverage for tag merge functionality

andersonjeccel committed:
feat: add success flash message when tags are merged

andersonjeccel committed:
feat: exclude current tag from merge options to prevent self-merge

andersonjeccel committed:
style: apply CS Fixer formatting to tag merge files

andersonjeccel committed:
fix: resolve PHPStan errors and apply CS Fixer to tests

andersonjeccel committed:
fix: remove unused property in TagModelTest

andersonjeccel committed:
style: apply CS Fixer formatting improvements

andersonjeccel committed:
refactor: improve mergeAction method to address feedback

- Reduce cognitive complexity from 22 to 15 by extracting helper methods
- Define constants for repeated permission literals
- Reduce return statements from 6 to 3 by consolidating logic
- Define URL constant for repeated merge path
- Fix namespace conflicts in controller

andersonjeccel committed:
fix: address remaining feedback issues

- Remove unused MERGE_URL_PATH constant
- Update all permission checks to use constants instead of hardcoded strings
- Remove unused  parameter from handleMergePostRequest
- Add PERMISSION_CREATE constant for consistency

andersonjeccel committed:
style: apply CS Fixer formatting improvements

andersonjeccel committed:
rector

andersonjeccel committed:
Fix TagController tests and improve code quality

- Fix unique constraint violation in tests by using unique usernames
- Add proper type declarations for objectId parameters in TagController methods
- Implement correct permission system for tag manager functionality
- Fix test data mismatch (tag names now match test expectations)
- Remove unused createAndLoginUser method to fix PHPStan issue
- Restructure test setup to avoid user creation conflicts
- Apply code quality improvements from Rector and CS Fixer
- Remove redundant PHPDoc tags that were redundant with method signatures
- Fix code style issues (spacing, imports, etc.)

All 17 TagController tests now pass successfully.

andersonjeccel committed:
Merge branch '7.x' into UXUI-155-merge-tags

kingsedem reviewed (approved):
Tested and works as expected

<img width="974" height="404" alt="image" src="https://github.com/user-attachments/assets/41a4feb6-44bb-4478-8889-12d8c7b34c03" />


<img width="933" height="301" alt="image" src="https://github.com/user-attachments/assets/3a69c218-e01d-4c61-a8ff-810c86107eac" />

andersonjeccel committed:
Merge branch '7.x' into UXUI-155-merge-tags

shinde-rahul commented on plugins/MauticTagManagerBundle/Controller/TagController.php:
```diff
@@ -485,6 +485,232 @@ public function viewAction(Request $request, TagDependencies $tagDependencies, $
+    /**
+     * Merge two tags together.
+     */
+    public function mergeAction(Request $request, int $objectId): Response
```
```suggestion
    public function mergeAction(Request $request, int $objectId, TagModel $model): Response
```
As we have autowiring, we should use DI.

shinde-rahul commented on plugins/MauticTagManagerBundle/Controller/TagController.php:
```diff
@@ -485,6 +485,232 @@ public function viewAction(Request $request, TagDependencies $tagDependencies, $
+        }
+
+        /** @var TagModel $model */
+        $model        = $this->getModel('lead.tag');
```
```suggestion
```

andersonjeccel committed:
Update plugins/MauticTagManagerBundle/Controller/TagController.php

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

andersonjeccel committed:
Update plugins/MauticTagManagerBundle/Controller/TagController.php

Co-authored-by: Rahul Shinde <shinde.r.a@gmail.com>

andersonjeccel committed:
Merge branch '7.x' into UXUI-155-merge-tags

andersonjeccel committed:
improve code quality

shinde-rahul commented on plugins/MauticTagManagerBundle/Controller/TagController.php:
```diff
@@ -151,140 +159,147 @@ public function indexAction(Request $request, $page = 1)
-                    }
-                }
-            }
+        if (!$this->security->isGranted(self::PERMISSION_CREATE)) {
```
Prefer early returns instead of assigning `$this->accessDenied()` to a variable and wrapping the rest of the logic in an else block.
This improves readability and follows common Mautic controller patterns.

shinde-rahul commented on app/bundles/LeadBundle/Tests/Model/TagModelTest.php:
```diff
@@ -0,0 +1,62 @@
+        $result = $tagModel->tagMerge($tag, $tag);
+
+        $this->assertSame($tag, $result);
+    }
```
This test mocks `TagModel::tagMerge()` itself, which makes the test ineffective.
This test does not validate any behavior and should either use a real `TagModel` instance or be removed.

shinde-rahul commented on app/bundles/LeadBundle/Tests/Model/TagModelTest.php:
```diff
@@ -0,0 +1,62 @@
+        $this->assertSame($tag, $result);
+    }
+
+    public function testTagMergeEvent(): void
```
Let's move it to `TagMergeEventTest`, since this isn't testing the integration with `TagModel`.

shinde-rahul commented on plugins/MauticTagManagerBundle/Form/Type/TagMergeType.php:
```diff
@@ -0,0 +1,62 @@
+<?php
+
+namespace MauticPlugin\MauticTagManagerBundle\Form\Type;
```
```suggestion
<?php

declare(strict_types=1);

namespace MauticPlugin\MauticTagManagerBundle\Form\Type;
```

shinde-rahul commented on plugins/MauticTagManagerBundle/Resources/views/Tag/merge.html.twig:
```diff
@@ -0,0 +1,5 @@
+{{ form_start(form) }}
+    <div class="hide">
```
Please include a brief comment explaining the use of hide, as it relies on AJAX behavior. This will reduce confusion and enhance maintainability for future contributors.

shinde-rahul reviewed (changes requested):
Overall LGTM 👍

I’ve added a few review comments. Once those are resolved, I’m good with this PR.

andersonjeccel committed:
Merge branch '7.x' into UXUI-155-merge-tags

andersonjeccel committed:
Address feedback

andersonjeccel committed:
fix: tests

patrykgruszka reviewed (changes requested):
Shouldn't we also update all usages of the removed tag? Removing it could break the functionality of:
- campaigns
- segments
- form actions
- point actions
- reports

andersonjeccel committed:
merge upstream 7.x into tag merge

andersonjeccel committed:
update merged tag references

andersonjeccel committed:
Merge branch '7.x' into UXUI-155-merge-tags

andersonjeccel committed:
fix tag manager sonar issues

andersonjeccel committed:
Merge branch '7.x' into UXUI-155-merge-tags

andersonjeccel committed:
Merge branch '7.x' into UXUI-155-merge-tags

patrykgruszka commented on app/bundles/LeadBundle/Model/TagModel.php:
```diff
@@ -100,4 +115,245 @@ protected function dispatchEvent($action, &$entity, $isNew = false, ?Event $even
+
+        $this->dispatcher->dispatch($event, LeadEvents::TAG_POST_MERGE);
+
+        $this->deleteEntity($secTag);
```
No transaction / atomicity around the merge. `tagMerge` performs many separate flushes - addTagsToLeads/removeTagsFromLeads flush per lead, then segment/report/campaign persists, then saveEntity, then deleteEntity. If anything fails partway (e.g. a lead with 50k associations, a constraint error, a timeout), you're left with a half-merged, inconsistent state: some leads moved, dependencies partially rewritten, secondary tag still present. A merge is a destructive, multi-table operation - it should be wrapped in $this->em->getConnection()->transactional(...) (or beginTransaction/commit/rollback) so it's all-or-nothing.

patrykgruszka commented on app/bundles/LeadBundle/Model/TagModel.php:
```diff
@@ -100,4 +115,245 @@ protected function dispatchEvent($action, &$entity, $isNew = false, ?Event $even
         return null;
     }
+
+    public function tagMerge(Tag $mainTag, Tag $secTag): Tag
```
victor/loser naming in TagMergeEvent is informal relative to the mainTag/secTag (and controller primaryTag/secondaryTag) used everywhere else. Three different vocabularies for the same two things across the PR. Aligning on one naming would reduce friction.

patrykgruszka commented on plugins/MauticTagManagerBundle/Controller/TagController.php:
```diff
@@ -156,123 +164,131 @@ public function newAction(Request $request, TagDependencies $tagDependencies)
+    /**
+     * @param array<string, mixed> $postActionVars
+     */
+    private function createTagModifyResponse(Request $request, Tag $tag, TagDependencies $tagDependencies, array $postActionVars, string $action, bool $ignorePost): Response
```
Duplicated docblock for this method

patrykgruszka commented on plugins/MauticTagManagerBundle/Controller/TagController.php:
```diff
@@ -390,10 +423,11 @@ private function getTag($tagId)
+    /**
+     * @return array<string, mixed>
+     */
+    private function getPostActionVars(Request $request, ?int $objectId = null): array
```
Duplicated docblock

patrykgruszka commented on app/bundles/LeadBundle/Model/TagModel.php:
```diff
@@ -100,4 +115,245 @@ protected function dispatchEvent($action, &$entity, $isNew = false, ?Event $even
+        $this->replaceMergedTagReferences($mainTag, $secTag);
+
+        $event = new TagMergeEvent($mainTag, $secTag);
+        $this->dispatcher->dispatch($event, LeadEvents::TAG_PRE_MERGE);
```
TAG_PRE_MERGE fires too late to be a true "pre" hook. The lead re-association and replaceMergedTagReferences both run before TAG_PRE_MERGE is dispatched. A listener on TAG_PRE_MERGE documented as "dispatched before two tags are merged" actually fires after the bulk of the merge has already happened. If the intent is to let listeners veto or observe the pre-merge state, dispatch it first - before any mutation

patrykgruszka commented on plugins/MauticTagManagerBundle/Translations/en_US/messages.ini:
```diff
@@ -17,4 +17,7 @@ mautic.protip.tags.multivalue="Use tags instead of custom fields for multi-value
+mautic.tagmanager.contentblock.copy="Use tags like keywords or sticky notes to mark interests (\"product-A-interest\"), track actions (\"attended-webinar-X\"), note preferences (\"prefers-video\"), or manage temporary statuses."
+mautic.tagmanager.tag.header.merge="Merging current tag"
+mautic.tagmanager.to.merge.into="Choose a tag to merge into"
+mautic.tagmanager.tag.merge.success="Tag '%secondary%' has been successfully merged into '%primary%'"
```
`mautic.tagmanager.tag.merge.success` is not used anywhere

patrykgruszka commented on app/bundles/LeadBundle/Model/TagModel.php:
```diff
@@ -100,4 +115,245 @@ protected function dispatchEvent($action, &$entity, $isNew = false, ?Event $even
+            $repo = $this->getRepository();
+            $repo->addTagsToLeads($leadIds, [$mainTag->getId()]);
+            $repo->removeTagsFromLeads($leadIds, [$secTag->getId()]);
+        }
```
Performance: lead re-association hydrates and flushes per contact.

tagMerge re-points contacts via addTagsToLeads/removeTagsFromLeads → updateTagsInLeads (TagRepository.php:165), which find()s every Lead and flush()es inside the loop (TagRepository.php:190). A tag on N contacts means N hydrations and ~2N flushes synchronously - this will OOM or time out on large tags.

The association is a plain join table (lead_tags_xref, no entity), so two set-based statements replace the whole loop:
```php
$conn = $this->em->getConnection();
$prefix = MAUTIC_TABLE_PREFIX;

// UPDATE IGNORE skips leads already carrying the primary tag (dup-key on lead_id+tag_id)
$conn->executeStatement(
    "UPDATE IGNORE {$prefix}lead_tags_xref SET tag_id = :main WHERE tag_id = :sec",
    ['main' => $mainTag->getId(), 'sec' => $secTag->getId()]
);
// clean up rows that couldn't move (lead had both tags)
$conn->executeStatement(
    "DELETE FROM {$prefix}lead_tags_xref WHERE tag_id = :sec",
    ['sec' => $secTag->getId()]
);
```
O(1) round-trips, no entities in memory, and handles the both-tags dedup case the current path covers implicitly. The $leadIds pre-fetch at TagModel.php:128-134 becomes unnecessary.

andersonjeccel committed:
merge 7.x into uxui-155 merge tags

# Conflicts:
#	plugins/MauticTagManagerBundle/Tests/Functional/Controller/TagControllerTest.php

andersonjeccel committed:
fix tag merge test assertions

andersonjeccel committed:
address tag merge review

andersonjeccel committed:
Merge branch '7.x' into UXUI-155-merge-tags

patrykgruszka reviewed (approved):
Code looks fine and works as described 👍

andersonjeccel committed:
Merge branch '7.x' into UXUI-155-merge-tags

matbcvo committed:
Merge branch '7.x' into UXUI-155-merge-tags
